### ISSUE TITLE
Model Predictions Were Narrow and Static Across Epochs

### ISSUE THREAD

**Issue by azadeh-gh** [NOAA-EMC/ocelot#29]
The model predictions had extremely low variance across all epochs. This issue was traced to missing gradient flow from the loss function back to the encoded observation features. After extensive debugging, The following changes were implemented to restore proper backpropagation and improve predictive dynamics.

### Fix Summary:
**1. Preserved Gradient Flow Through Encoding**

- Replaced all `.detach()` operations that inadvertently cut off gradient flow.
- Verified that `encoded` and `x_hidden` tensors had `requires_grad=True`.
- Used `retain_grad()` on key tensors (`encoded`, `x_hidden`) to inspect backpropagation flow.

**2. Refactored forward() for Clarity & Gradient Tracking**

- Added logic to conditionally use `scatter_mean` or `index_add` with zero-padding.
- Ensured all tensor operations preserved gradient flow:
```python
x_hidden = self.aggregate_encoded(encoded, tgt_encoder, num_targets)
```
**3. Added debug() Logging Function with verbose Flag**

- Wrapped debug prints to control verbosity across all ranks:
```python
def debug(self, *args, **kwargs):
    if self.verbose and (not hasattr(self, "trainer") or self.trainer.is_global_zero):
        print(*args, **kwargs)
```

- Activated via CLI flag:

```python
python train_gnn.py --verbose
```
**4. Updated `training_step` and `on_after_backward` for Probes**

- Probed gradients via torch.autograd.grad():
```python
probe = grad(self._x_hidden_ref.sum(), self._encoded_ref, retain_graph=True, allow_unused=True)[0]
```
**5. Added Support for instrument_ids in Data Pipeline**

- Modified `process_timeseries`, `GNNDataModule`, and `BinDataset` to pass `instrument_ids`.
- Updated model to compute weighted loss via `ocelot_loss()`.

### Result

- Predictions are now dynamic and spread across the expected range.
- Gradients correctly flow from loss → x_out → x_hidden → encoded → obs features.
- Logs confirm non-zero gradient norms at each stage.
